File: ios/SOTALog/Services/Database/SyncImporter.swift
import Foundation

enum SyncImporter {

    // MARK: - Types

    struct ActivationKey: Hashable {
        let date: String
        let potaReference: String?
        let sotaReference: String?

        // Carried for Log creation but NOT part of equality/hash
        var stationCallsign: String
        var myGrid: String?

        func hash(into hasher: inout Hasher) {
            hasher.combine(date)
            hasher.combine(potaReference)
            hasher.combine(sotaReference)
        }

        static func == (lhs: ActivationKey, rhs: ActivationKey) -> Bool {
            lhs.date == rhs.date &&
            lhs.potaReference == rhs.potaReference &&
            lhs.sotaReference == rhs.sotaReference
        }
    }

    struct ParsedQSORecord {
        var qso: QSO
        var rawFields: [String: String]
    }

    struct GroupingResult {
        var activations: [(key: ActivationKey, qsos: [ParsedQSORecord])]
        var unattached: [ParsedQSORecord]
    }

    // MARK: - Grouping

    static func groupByActivation(
        records: [[String: String]],
        fallbackCallsign: String?,
        validPotaRefs: [String: String],
        validSotaCodes: [String: String]
    ) -> GroupingResult {
        var groups: [ActivationKey: [ParsedQSORecord]] = [:]
        var unattached: [ParsedQSORecord] = []

        for fields in records {
            guard let qso = ADIFFormatter.qsoFromFields(fields) else { continue }
            let record = ParsedQSORecord(qso: qso, rawFields: fields)

            // Extract and validate POTA reference
            let rawPota = fields["MY_SIG_INFO"]
            let validatedPota: String? = rawPota.flatMap { raw in
                let normalized = POTAPark.normalize(raw)
                return validPotaRefs[normalized]
            }

            // Extract and validate SOTA reference
            let rawSota = fields["MY_SOTA_REF"]
            let validatedSota: String? = rawSota.flatMap { raw in
                let normalized = SOTASummit.normalize(raw)
                return validSotaCodes[normalized]
            }

            // Station callsign with fallback
            let stationCallsign = fields["STATION_CALLSIGN"] ?? fallbackCallsign ?? ""

            // Grid square
            let myGrid = fields["MY_GRIDSQUARE"]

            if validatedPota == nil && validatedSota == nil {
                unattached.append(record)
            } else {
                let key = ActivationKey(
                    date: qso.date,
                    potaReference: validatedPota,
                    sotaReference: validatedSota,
                    stationCallsign: stationCallsign,
                    myGrid: myGrid
                )
                groups[key, default: []].append(record)
            }
        }

        // Stable ordering by date then reference
        let sorted = groups.sorted { a, b in
            if a.key.date != b.key.date { return a.key.date < b.key.date }
            let aRef = a.key.potaReference ?? a.key.sotaReference ?? ""
            let bRef = b.key.potaReference ?? b.key.sotaReference ?? ""
            return aRef < bRef
        }

        return GroupingResult(
            activations: sorted.map { (key: $0.key, qsos: $0.value) },
            unattached: unattached
        )
    }
}

File: android/app/src/main/kotlin/com/sotalog/android/domain/services/SyncImporter.kt
package com.sotalog.android.domain.services

import com.sotalog.android.domain.models.POTAPark
import com.sotalog.android.domain.models.QSO
import com.sotalog.android.domain.models.SOTASummit

object SyncImporter {

    data class ActivationKey(
        val date: String,
        val potaReference: String?,
        val sotaReference: String?,
        val stationCallsign: String,
        val myGrid: String?,
    ) {
        // Equality and hash based on date + potaReference + sotaReference only
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ActivationKey) return false
            return date == other.date &&
                potaReference == other.potaReference &&
                sotaReference == other.sotaReference
        }

        override fun hashCode(): Int {
            var result = date.hashCode()
            result = 31 * result + (potaReference?.hashCode() ?: 0)
            result = 31 * result + (sotaReference?.hashCode() ?: 0)
            return result
        }
    }

    data class ParsedQSORecord(
        val qso: QSO,
        val rawFields: Map<String, String>,
    )

    data class GroupingResult(
        val activations: List<Pair<ActivationKey, List<ParsedQSORecord>>>,
        val unattached: List<ParsedQSORecord>,
    )

    /**
     * Groups parsed ADIF records into activations by date and reference.
     *
     * @param records parsed ADIF field maps
     * @param fallbackCallsign callsign to use when STATION_CALLSIGN is absent
     * @param validPotaRefs map of normalized POTA ref -> canonical ref
     * @param validSotaCodes map of normalized SOTA code -> canonical code
     */
    fun groupByActivation(
        records: List<Map<String, String>>,
        fallbackCallsign: String?,
        validPotaRefs: Map<String, String>,
        validSotaCodes: Map<String, String>,
    ): GroupingResult {
        val groups = linkedMapOf<ActivationKey, MutableList<ParsedQSORecord>>()
        val unattached = mutableListOf<ParsedQSORecord>()

        for (fields in records) {
            val qso = ADIFFormatter.qsoFromFields(fields) ?: continue
            val record = ParsedQSORecord(qso, fields)

            // Extract and validate POTA reference
            val rawPota = fields["MY_SIG_INFO"]
            val validatedPota = rawPota?.let { raw ->
                val normalized = POTAPark.normalize(raw)
                validPotaRefs[normalized]
            }

            // Extract and validate SOTA reference
            val rawSota = fields["MY_SOTA_REF"]
            val validatedSota = rawSota?.let { raw ->
                val normalized = SOTASummit.normalize(raw)
                validSotaCodes[normalized]
            }

            // Station callsign with fallback
            val stationCallsign = fields["STATION_CALLSIGN"] ?: fallbackCallsign ?: ""

            // Grid square
            val myGrid = fields["MY_GRIDSQUARE"]

            if (validatedPota == null && validatedSota == null) {
                unattached += record
            } else {
                val key = ActivationKey(
                    date = qso.date,
                    potaReference = validatedPota,
                    sotaReference = validatedSota,
                    stationCallsign = stationCallsign,
                    myGrid = myGrid,
                )
                groups.getOrPut(key) { mutableListOf() } += record
            }
        }

        // Stable ordering by date then reference
        val sorted = groups.entries.sortedWith(compareBy(
            { it.key.date },
            { it.key.potaReference ?: it.key.sotaReference ?: "" },
        ))

        return GroupingResult(
            activations = sorted.map { it.key to it.value.toList() },
            unattached = unattached,
        )
    }
}
